src/Microsoft.AspNetCore.Razor.Evolution/RazorSourceDocument.cs

```diff
@@ -50,31 +50,38 @@ public static RazorSourceDocument ReadFrom(Stream stream, string filename, Encod
 
         private static RazorSourceDocument ReadFromInternal(Stream stream, string filename, Encoding encoding)
         {
-            var reader = new StreamReader(
-                stream,
-                encoding ?? Encoding.UTF8,
-                detectEncodingFromByteOrderMarks: true,
-                bufferSize: (int)stream.Length,
-                leaveOpen: true);
-
-            using (reader)
-            {
-                var content = reader.ReadToEnd();
+            var streamLength = (int)stream.Length;
+            var content = string.Empty;
+            var contentEncoding = encoding ?? Encoding.UTF8;
 
-                if (encoding == null)
-                {
-                    encoding = reader.CurrentEncoding;
-                }
-                else if (encoding != reader.CurrentEncoding)
+            if (streamLength > 0)
+            {
+                var reader = new StreamReader(
+                    stream,
+                    contentEncoding,
+                    detectEncodingFromByteOrderMarks: true,
+                    bufferSize: streamLength,
```

Comment:
We could probably just solve this issue as well with a `4096` here. This code also has the problem that it will try to read the buffer in one whole go which is wicked bad.